### ISSUE TITLE
Feature: Learning subpackage suggestions for 1.0

### ISSUE THREAD

**Issue by rushilg99** [macrosynergy/macrosynergy#1797]
My suggestions for 1.0 of the package, specifically for learning:

NB: if ticked, it means either it is complete or there is a good reason (which is provided) why this cannot go in for 1.0

- [x] Abstract base class for the learning process. BetaEstimator and SignalOptimizer should inherit from this.
- [x] "Metric" renamed to "scorer"
- [x] Get rid of GridSearchCV in the underlying code and write our own hyperparameter search method that allows for a grid search, random search and Bayesian hyperparameter search. (**VETOED for 1.0**)
- [x] If a list of scorers is provided, each model and hyperparameter candidate is evaluated based on each of the metrics. The metric for each model is adjusted by (score - min_score)/(max_score - min_score) based on the largest and smallest metric across models. These standardised scores are averaged across metrics and the model with highest composite score is chosen. 
- [x] Option to include the standard deviation across splits in the score calculation. For a single metric, we would compute negative standard deviation across splits and adjust by comparing with competing models/hyperparameters, as per the calculation in the previous bullet, before creating a composite score by averaging mean and negative standard deviation - the one with the best score is selected. If multiple metrics are specified, then these composite scores are computed for each metric, following which the procedure in the previous bullet point takes over. 
- [x] If multiple splitters are passed, then the final cross-validation score is computed for each splitter and the average is returned. 
- [x] A new splitter for which n_periods number of native dataset frequency periods are specified and n_splits are specified. The last n_splits * n_periods are divided into n_splits different test sets at the END of the training history, resulting in an expanding window cross-validator with a specified validation set length that focuses on the most recent samples. A special case of n_splits = 1 results in a single validation set. Come up with a good name for this. 
- [x] Store optimal cross-validation/validation scores at each rebalancing date. If multiple metrics were passed in, then the composite score, as well each realised average metric should be stored. This means that different pipelines can be compared without looking at a PnL.
- [x] Write class to contain the logic for adapting a sklearn metric to either, within a method, create a scorer over the panel (dropping legitimate NAs) or evaluate a score over each cross-section before returning the (possibly negative) mean/median.
- [x] Replace current implementation of the sharpe and sortino metrics with a better version that accounts for size of the signal in the case of regression.
- [x] The inner splitter should be specified in either calculate_predictions() or estimate_beta(), not in the constructor. 
- [x] Signal optimizer and beta estimator need to have methods to visualise average feature coefficients (because the systems of regressors have a coefs_ attribute not coef_) **(Not a priority)**
- [x] Replace "M" frequency with "ME" in all regressor systems **(Issue with Python 3.7 and athena if we do this now)**
- [x] LADRegression - remove restriction on requiring panel format data. i.e. doesn't need double indexed format
- [x] Make sure stored quantamental dataframes are memory efficient using the categorical dtype.
- [x] 90%+ code coverage for learning. 
- [x] Add option to pass scaler_y **(NOT NEEDED due to sklearn's transformedtargetregressor class)**
- [x] FGLS and Tail weighted least squares **(VETOED FOR 1.0)**
- [x] PanelPCA transformer
- [x] LARS selector
- [x] Review of the metrics we provide. Can provide better implementations of Sharpe & Sortino. Makes sense to add weighted metrics. 
- [x] For each transformer, predictor, selector we need an integration test with gridsearchcv and signaloptimizer **(TODO in a later release - not urgent)**
- [x] Review of splitters and reorganisation
- [x] Review of metrics - rename to model_evaluation - and reorganisation. 
- [x] Review of predictors. 

I will add more suggestions as I remember them/think of them.

**Comment by rushilg99** [comment]
For Eric

- [ ] n_splits storage for explainability in base_panel_learner
- [ ] n_splits timeplot refactor that shows the number of splits used for each splitter passed into run() for a particular pipeline. 
- [ ] Optimize base_panel_learner to make as fast as you can
- [ ] Check edge cases, particularly when n_metrics = n_splitters = 1. Does it yield the same output as previous signal optimizer and beta estimator classes? 

I'll add more when I remember, but start with these.

**Comment by rushilg99** [comment]
Remaining TODOs:

- [ ] Review of docs
- [ ] N splits time plot 
- [ ] Categorical dataframe
